File: leetcode/go/1562.find-latest-group-of-size-m.go
func findLatestStep(arr []int, m int) int {
    if len(arr) == m {
        return m;
    }
    ranges := make([]int, len(arr) + 2)
    step := 0
    ans := -1
    start := 0
    end := 0
    for _, i := range arr {
        if ranges[i-1] == m || ranges [i+1] == m {
            ans = step
        }
        step++
        start = i - ranges[i-1]
        end = i + ranges[i+1]
        ranges[start] = end - start + 1
        ranges[end] = end - start + 1
    }
    return ans
}

File: leetcode/rust-leetcode/src/1562.find-latest-group-of-size-m.rs
impl Solution {
    pub fn find_latest_step(arr: Vec<i32>, m: i32) -> i32 {
        let m = m as usize;
        if m == arr.len() {return m as i32;}
        let arr = arr.iter().map(|&n| n as usize).collect::<Vec<_>>();
        let mut ranges = vec![0; arr.len() + 2];
        let mut step = 0;
        let mut ans = -1;
        let (mut end, mut start) = (0, 0);
        for &n in arr.iter() {
            if ranges[n - 1] == m || ranges[n + 1] == m {
                ans = step;
            }
            step += 1;
            end = n + ranges[n + 1];
            start = n - ranges[n - 1];
            ranges[start] = end - start + 1;
            ranges[end] = end - start + 1;
            // println!("{:?}", ranges);
        }
        ans
    }
}

